crates/iceberg/src/writer/base_writer/equality_delete_writer.rs

```diff
@@ -0,0 +1,438 @@
+                res.partition(self.partition_value.clone());
+                res.build().expect("msg")
+            })
+            .collect_vec())
+    }
+}
+
+/// Help to project specific field from `RecordBatch`` according to the column id.
+pub struct FieldProjector {
+    index_vec_vec: Vec<Vec<usize>>,
+}
+
+impl FieldProjector {
+    /// Init FieldProjector
+    pub fn new(
+        batch_fields: &Fields,
+        column_ids: &[usize],
+        column_id_meta_key: &str,
+    ) -> Result<(Self, Fields)> {
+        let mut index_vec_vec = Vec::with_capacity(column_ids.len());
+        let mut fields = Vec::with_capacity(column_ids.len());
+        for &id in column_ids {
+            let mut index_vec = vec![];
+            if let Some(field) = Self::fetch_column_index(
+                batch_fields,
+                &mut index_vec,
+                id as i64,
+                column_id_meta_key,
+            ) {
+                fields.push(field.clone());
+                index_vec_vec.push(index_vec);
+            } else {
+                return Err(Error::new(
+                    ErrorKind::DataInvalid,
+                    format!("Can't find source column id: {}", id),
+                ));
+            }
+        }
+        Ok((Self { index_vec_vec }, Fields::from_iter(fields)))
+    }
+
+    fn fetch_column_index(
+        fields: &Fields,
+        index_vec: &mut Vec<usize>,
+        col_id: i64,
+        column_id_meta_key: &str,
+    ) -> Option<FieldRef> {
+        for (pos, field) in fields.iter().enumerate() {
+            let id: i64 = field
+                .metadata()
+                .get(column_id_meta_key)
+                .expect("column_id must be set")
+                .parse()
+                .expect("column_id must can be parse as i64");
+            if col_id == id {
+                index_vec.push(pos);
+                return Some(field.clone());
+            }
+            if let DataType::Struct(inner) = field.data_type() {
+                let res = Self::fetch_column_index(inner, index_vec, col_id, column_id_meta_key);
+                if !index_vec.is_empty() {
+                    index_vec.push(pos);
+                    return res;
+                }
+            }
+        }
+        None
+    }
+    /// Do projection with batch
+    pub fn project(&self, batch: &[ArrayRef]) -> Vec<ArrayRef> {
+        self.index_vec_vec
+            .iter()
+            .map(|index_vec| Self::get_column_by_index_vec(batch, index_vec))
+            .collect_vec()
+    }
+
+    fn get_column_by_index_vec(batch: &[ArrayRef], index_vec: &[usize]) -> ArrayRef {
+        let mut rev_iterator = index_vec.iter().rev();
+        let mut array = batch[*rev_iterator.next().unwrap()].clone();
+        for idx in rev_iterator {
+            array = array
+                .as_any()
+                .downcast_ref::<StructArray>()
+                .unwrap()
```

Comment:
May need to return a `Result` here and in `project` so that we can avoid the `unwrap`